staging/src/k8s.io/apimachinery/pkg/util/diff/diff.go

```diff
@@ -17,122 +20,53 @@ limitations under the License.
 package diff
 
 import (
-	"bytes"
+	"encoding/json"
 	"fmt"
 	"reflect"
-	"strings"
-	"text/tabwriter"
 
-	"github.com/google/go-cmp/cmp" //nolint:depguard
-	"k8s.io/apimachinery/pkg/util/dump"
+	"github.com/pmezard/go-difflib/difflib"
 )
 
-func legacyDiff(a, b interface{}) string {
-	return cmp.Diff(a, b)
-}
-
-// StringDiff diffs a and b and returns a human readable diff.
-// DEPRECATED: use github.com/google/go-cmp/cmp.Diff
-func StringDiff(a, b string) string {
-	return legacyDiff(a, b)
-}
+// Diff returns a string representation of the difference between two objects.
+// When built with the nogocmp tag, it uses difflib to generate a unified diff
+// of the JSON representation of the objects.
+func Diff(a, b any) string {
+	// Special handling for complex numbers which don't marshal to JSON properly
+	if isComplex(a) || isComplex(b) {
```

Comment:
I'm sort of surprised to see a single special case here for complex numbers ... what do they do that makes us have to handle them this way? are there really no other problematic inputs for json marshaling?

Reply:
removed `isComplex`